llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -612,6 +612,118 @@ Instruction *emitSpecConstantRecursive(Type *Ty, Instruction *InsertBefore,
 
 } // namespace
 
+/// Function creates load instruction from the given Buffer by the given Offset.
+/// Function returns the value of load instruction.
+static Value *createLoadFromBuffer(CallInst *CI, Value *Buffer, size_t Offset,
+                                   Type *SCType) {
+  LLVMContext &C = CI->getContext();
+  Type *Int8Ty = Type::getInt8Ty(C);
+  Type *Int32Ty = Type::getInt32Ty(C);
+  GetElementPtrInst *GEP = GetElementPtrInst::Create(
+      Int8Ty, Buffer, {ConstantInt::get(Int32Ty, Offset, false)}, "gep", CI);
+
+  Instruction *BitCast = nullptr;
+  if (SCType->isIntegerTy(1)) // No bitcast to i1 before load
+    BitCast = GEP;
+  else
+    BitCast = new BitCastInst(
+        GEP, PointerType::get(SCType, GEP->getAddressSpace()), "bc", CI);
+
+  // When we encounter i1 spec constant, we still load the whole byte
+  Value *Load = new LoadInst(SCType->isIntegerTy(1) ? Int8Ty : SCType, BitCast,
+                             "load", CI);
+  if (SCType->isIntegerTy(1)) // trunc back to i1 if necessary
+    Load = CastInst::CreateIntegerCast(Load, SCType, /* IsSigned */ false,
+                                       "tobool", CI);
+
+  return Load;
+}
+
+/// Function tries to dig out the initializer from the given CallInst to
+/// SpecConst function. ArgIndex is the expected index of the function operand
+/// leading to the initializer.
+///
+/// Examples:
+/// 1)
+///   %"spec_id" = type { i32 }
+///   @value = internal addrspace(1) constant %"spec_id" { i32 123 }, align 4
+///   call spir_func i32 @sycl_getScalar2020SpecConst(%1, @value, %2)
+///
+/// 2)
+///   %"spec_id" = type { %A }
+///   %A = type { i32 }
+///   @value = constant %"spec_id" { %A { i32 1 } }, align 4
+///   call spir_func void @getCompositeSpecConst(%1, %2, @value, %3)
+static Constant *tryGetSpecConstInitializerFromCI(CallInst *CI,
+                                                  unsigned ArgIndex) {
+  assert(CI->getNumOperands() > ArgIndex && "ArgIndex is out of boundary");
+  auto *GV = dyn_cast<GlobalVariable>(
+      CI->getArgOperand(ArgIndex)->stripPointerCasts());
+  if (!GV)
+    return nullptr;
+
+  // Go through global variable if the argument was not null.
+  assert(GV->hasInitializer() && "GV is expected to have initializer");
+  Constant *Initializer = GV->getInitializer();
+  assert((isa<ConstantAggregate>(Initializer) || Initializer->isZeroValue()) &&
+         "expected specialization_id instance");
+  // specialization_id structure contains a single field which is the
+  // default value of corresponding specialization constant.
+  return Initializer->getAggregateElement(0u);
+}
+
+/// Function replaces last Metadata node in the given vector with new
+/// node which contains given Padding.
+static void
+updatePaddingInLastMDNode(LLVMContext &Ctx,
```

Comment:
Unrelated to the PR, just an observation: we have at least one more place where we insert some padding metadata:

https://github.com/intel/llvm/blob/61c7c6449e6234cef56ad26e5cdbc7262f7f2f98/llvm/tools/sycl-post-link/SpecConstants.cpp#L252-L262

These paddings are hard and its been a while since I looked into this code. It would be good to refresh that knowledge to see if we could unify adding those paddings or at least add more explanatory comments, because right now I'm not sure that I understand why do we have both those places where we add special descriptors